sql/select.go

```diff
@@ -85,3 +99,509 @@ func (p *planner) expandSubqueries(stmt parser.Statement) error {
+		left := make(qvalueRangeMap)
+		left.analyzeExpr(t.Left)
+		right := make(qvalueRangeMap)
+		right.analyzeExpr(t.Right)
+
+		for id, lRange := range left {
+			rRange, ok := right[id]
+			if !ok {
+				continue
+			}
+			r := &qvalueRange{}
+			if lRange.start.datum != nil && rRange.start.datum != nil {
+				r.start = lRange.start
+				r.start.union(rRange.start, true)
+			}
+			if lRange.end.datum != nil && rRange.end.datum != nil {
+				r.end = lRange.end
+				r.end.union(rRange.end, false)
+			}
+			if r.start.datum != nil || r.end.datum != nil {
+				m[id] = r
+			}
+		}
+
+	case *parser.AndExpr:
+		m.analyzeExpr(t.Left)
+		m.analyzeExpr(t.Right)
+
+	case *parser.ParenExpr:
+		m.analyzeExpr(t.Expr)
+
+	case *parser.ComparisonExpr:
+		m.analyzeComparisonExpr(t)
+
+	case *parser.RangeCond:
+		if t.Not {
+			// "a NOT BETWEEN b AND c" -> "a < b OR a > c"
+			m.analyzeExpr(&parser.OrExpr{
+				Left: &parser.ComparisonExpr{
+					Operator: parser.LT,
+					Left:     t.Left,
+					Right:    t.From,
+				},
+				Right: &parser.ComparisonExpr{
+					Operator: parser.GT,
+					Left:     t.Left,
+					Right:    t.To,
+				},
+			})
+		} else {
+			// "a BETWEEN b AND c" -> "a >= b AND a <= c"
+			m.analyzeExpr(&parser.AndExpr{
+				Left: &parser.ComparisonExpr{
+					Operator: parser.GE,
+					Left:     t.Left,
+					Right:    t.From,
+				},
+				Right: &parser.ComparisonExpr{
+					Operator: parser.LE,
+					Left:     t.Left,
+					Right:    t.To,
+				},
+			})
+		}
+	}
+}
+
+// analyzeComparison analyzes the comparison expression, restricting the start
+// and end info in indexInfo.elems if possible.
+func (m qvalueRangeMap) analyzeComparisonExpr(node *parser.ComparisonExpr) {
+	op := node.Operator
+	switch op {
+	case parser.EQ, parser.LT, parser.LE, parser.GT, parser.GE:
+		break
+
+	default:
+		// TODO(pmattis): For parser.LIKE we could extract the constant prefix and
+		// treat as a range restriction.
+		return
+	}
+
+	qval, ok := node.Left.(*qvalue)
+	var constExpr parser.Expr
+	if ok {
+		if !isConst(node.Right) {
+			return
+		}
+		constExpr = node.Right
+		// qval <op> constant
+	} else {
+		qval, ok = node.Right.(*qvalue)
+		if !ok {
+			return
+		}
+		if !isConst(node.Left) {
+			return
+		}
+		constExpr = node.Left
+		// constant <op> qval: invert the operation.
+		switch op {
+		case parser.LT:
+			op = parser.GT
+		case parser.LE:
+			op = parser.GE
+		case parser.GT:
+			op = parser.LT
+		case parser.GE:
+			op = parser.LE
+		}
+	}
+
+	// Evaluate the constant expression.
+	datum, err := parser.EvalExpr(constExpr)
+	if err != nil {
+		return
+	}
+
+	// Ensure the resulting datum matches the column type.
+	_, err = convertDatum(qval.col, datum)
+	if err != nil {
+		return
+	}
+
+	if log.V(2) {
+		log.Infof("analyzeComparisonExpr: %s %s %s",
+			qval.col.Name, op, datum)
+	}
+
+	r := m.getRange(qval.col.ID)
+
+	// Restrict the start element.
+	switch startOp := op; startOp {
+	case parser.GE:
+		startOp = parser.EQ
```

Comment:
Yes. I've reworked this so that "start" endpoints require the op to be `GE` or `GT` and "end" endpoints require the op to be `LE` or `LT`.